librz/core/canalysis.c

```diff
@@ -6601,3 +6601,294 @@ RZ_API void rz_core_analysis_cc_init(RzCore *core) {
+
+RZ_API void rz_core_analysis_class_json(RzAnalysis *analysis, PJ *pj, const char *class_name) {
+	pj_o(pj);
+	pj_ks(pj, "name", class_name);
+
+	pj_k(pj, "bases");
+	pj_a(pj);
+	RzVector *bases = rz_analysis_class_base_get_all(analysis, class_name);
+	if (bases) {
+		RzAnalysisBaseClass *base;
+		rz_vector_foreach(bases, base) {
+			pj_o(pj);
+			pj_ks(pj, "id", base->id);
+			pj_ks(pj, "name", base->class_name);
+			pj_kn(pj, "offset", base->offset);
+			pj_end(pj);
+		}
+		rz_vector_free(bases);
+	}
+	pj_end(pj);
+
+	pj_k(pj, "vtables");
+	pj_a(pj);
+	RzVector *vtables = rz_analysis_class_vtable_get_all(analysis, class_name);
+	if (vtables) {
+		RzAnalysisVTable *vtable;
+		rz_vector_foreach(vtables, vtable) {
+			pj_o(pj);
+			pj_ks(pj, "id", vtable->id);
+			pj_kn(pj, "addr", vtable->addr);
+			pj_kn(pj, "offset", vtable->offset);
+			pj_end(pj);
+		}
+	}
+	pj_end(pj);
+
+	pj_k(pj, "methods");
+	pj_a(pj);
+	RzVector *methods = rz_analysis_class_method_get_all(analysis, class_name);
+	if (methods) {
+		char *method_type[] = { "DEFAULT", "VIRTUAL", "V_DESTRUCTOR", "DESTRUCTOR", "CONSTRUCTOR" };
+		RzAnalysisMethod *meth;
+		rz_vector_foreach(methods, meth) {
+			pj_o(pj);
+			pj_ks(pj, "name", meth->real_name);
+			pj_kn(pj, "addr", meth->addr);
+			pj_ks(pj, "type", method_type[meth->method_type]);
+			if (meth->vtable_offset >= 0) {
+				pj_kn(pj, "vtable_offset", (ut64)meth->vtable_offset);
+			}
+			pj_end(pj);
+		}
+		rz_vector_free(methods);
+	}
+	pj_end(pj);
+
+	pj_end(pj);
+}
+
+typedef struct {
+	RzAnalysis *analysis;
+	PJ *pj;
+} ListJsonCtx;
+
+static bool rz_core_analysis_class_list_json_cb(void *user, const char *k, const char *v) {
+	ListJsonCtx *ctx = user;
+	rz_core_analysis_class_json(ctx->analysis, ctx->pj, k);
+	return true;
+}
+
+static void rz_core_analysis_class_list_json(RzAnalysis *analysis, PJ *pj) {
+	ListJsonCtx ctx;
+	ctx.analysis = analysis;
+	ctx.pj = pj;
+	pj_a(pj);
+	rz_analysis_class_foreach(analysis, rz_core_analysis_class_list_json_cb, &ctx);
+	pj_end(pj);
+	return;
+}
+
+RZ_API void rz_core_analysis_class_list(RzAnalysis *analysis, RzCmdStateOutput *state) {
+	rz_return_if_fail(analysis && state);
+	if (state->mode == RZ_OUTPUT_MODE_JSON) {
+		rz_core_analysis_class_list_json(analysis, state->d.pj);
+		return;
+	}
+
+	SdbList *classes = rz_analysis_class_get_all(analysis, state->mode != RZ_OUTPUT_MODE_RIZIN);
+	SdbListIter *iter;
+	SdbKv *kv;
+	if (state->mode == RZ_OUTPUT_MODE_RIZIN) {
+		ls_foreach (classes, iter, kv) {
+			// need to create all classes first, so they can be referenced
+			rz_cons_printf("ac %s\n", sdbkv_key(kv));
+		}
+		ls_foreach (classes, iter, kv) {
+			rz_core_analysis_class_print_cmd(analysis, sdbkv_key(kv));
+		}
+	} else {
+		ls_foreach (classes, iter, kv) {
+			rz_core_analysis_class_print(analysis, sdbkv_key(kv), state->mode == RZ_OUTPUT_MODE_LONG);
+		}
+	}
+	ls_free(classes);
+}
+
+RZ_API void rz_core_analysis_class_list_bases(RzAnalysis *analysis, const char *class_name) {
```

Comment:
Same applies to all these functions. Moreover, I wonder if it is better to make them `RZ_IPI` instead, they don't make sense as the API for the use outside of Rizin core.

Reply:
Yep it was on my list.